### ISSUE TITLE
Sandstorm will restart its HTTPS server when it must, due to a cert expiring

### ISSUE THREAD

**Issue by paulproteus** [sandstorm-io/sandcats#85]

**Comment by paulproteus** [comment]
I used to think the best way to handle this is for Sandstorm to parse the certificate when it receives it and check the expiry.

Now I believe the best way to handle this is that Sandstorm should be told, when it receives a certificate, when it should renew it.

**Comment by paulproteus** [comment]
The design is that:
- When Sandstorm starts, it uses the newest cert it finds where the `notBefore` is now-or-the-future
- When Sandstorm asks for a new cert, it notices the `notBefore` of the new cert and stores it in a variable somewhere.
- In the `SNICallback`, Sandstorm checks if that variable is set. If so, it checks the current time on every inbound request. If that time is after the the time stored in that variable, it switches to the new cert, and unsets the variable.

This does mean that, for about one day per week, Sandstorm is going to call `gettimeofday` with every inbound HTTPS request, which is arguably kind of tragic.

This also means we need to change sandcats to provide this data.

## Relevant technical facts:

```
var notBefore = metadata['notBefore'];
```

in https://github.com/sandstorm-io/sandstorm/pull/900/files

**Comment by paulproteus** [comment]
Other notes:
- We could handle `notAfter` but I figure it's OK to just keep using an expired certificate if we really have no better cert to use.
- We can avoid the `gettimeofday` call for every single inbound connection by setting a `recheckCertificates` server-side global variable (or some smarter design) when we fetch fresh certs, and then we set it to `false` when we've updated to the new cert. In fact, we could even be more specific -- `nextCertificateInfo` could be `null` normally, and when it's set to an object like `{filename: 5, notBefore: 1442858259}` we check the time, and then we set the cert to that one when the time comes, and we then set that variable to `null` so that we stop calling `gettimeofday` when we're not waiting for anything.
- This design means that non-SNI clients might sometimes receive an expired certificate. I think I can live with that.
- OTHER: It's probably smart to always have 2 certs, so that if sandcats goes down for a full week, users aren't hosed. In which case we'll always be doing the `gettimeofday` check anyway, oh well.
